HarmonyPatches/HitSoundPatches.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using System.Reflection.Emit;
+using HarmonyLib;
+using UnityEngine;
+
+namespace SongCore.HarmonyPatches
+{
+    [HarmonyPatch(typeof(NoteCutSoundEffectManager), nameof(NoteCutSoundEffectManager.HandleNoteWasSpawned))]
+    internal class NoteCutSoundEffectManagerHandleNoteWasSpawnedPatch
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions)
+        {
+            // Removes the condition that filters notes that are not in chronological order.
+            return new CodeMatcher(instructions)
+                .MatchForward(false, new CodeMatch(OpCodes.Ldloc_0),
+                    new CodeMatch(OpCodes.Callvirt),
+                    new CodeMatch(OpCodes.Brtrue))
+                .RemoveInstructions(22)
+                .InstructionEnumeration();
+        }
+    }
+
+    [HarmonyPatch(typeof(NoteCutSoundEffect), nameof(NoteCutSoundEffect.Init))]
+    internal class NoteCutSoundEffectInitPatch
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions)
+        {
+            // Removes the PlayScheduled call from the original method.
+            // Because it plays the sound instantly rather than after the passed delay.
+            return new CodeMatcher(instructions)
+                .MatchForward(false, new CodeMatch(OpCodes.Ldarg_0),
+                    new CodeMatch(OpCodes.Ldfld),
+                    new CodeMatch(OpCodes.Ldarg_0),
+                    new CodeMatch(OpCodes.Ldfld),
+                    new CodeMatch(OpCodes.Callvirt),
+                    new CodeMatch(OpCodes.Ret))
+                .RemoveInstructions(5)
+                .InstructionEnumeration();
+        }
+
+        private static void Postfix(NoteCutSoundEffect __instance, AudioSource ____audioSource, double ____startDSPTime)
+        {
+            __instance.StartCoroutine(PlayHitSoundCoroutine(____audioSource, ____startDSPTime));
+        }
+
+        private static IEnumerator PlayHitSoundCoroutine(AudioSource audioSource, double startDSPTime)
+        {
+            yield return new WaitUntil(() => AudioSettings.dspTime > startDSPTime);
+            audioSource.Play();
```

Comment:
If I'm not wrong this would make hitsounds only play during frames, which would be a regression from the base game. I highly recommend still scheduling playback ahead of time, but perhaps a shorter delay to mitigate the issue.

Reply:
> I highly recommend still scheduling playback ahead of time</p>

We can't rely on `audioSource.PlayScheduled` since it doesn't actually wait for the delay before *playing* the sound. So whenever you'll call it it will play the sound instantly, so that would not fix the issue. `WaitUntil` waits for frames, and `WaitForSeconds` isn't `double` precise (and waits for frames as well). I can't think of a better way to handle it currently than using a coroutine, but I am well open to suggestions.

The best thing I can think of is still using it, but later than when notes spawn.

```cs
// Schedule it 250 ms in advance so we're not frame-dependant.
yield return new WaitUntil(() => startDSPTime - AudioSettings.dspTime <= 0.025);
audioSource.PlayScheduled(startDSPTime);
```

Or a lower value if it is still reliable.

Reply:
> The best thing I can think of is still using it, but later than when notes spawn:
> 
> ```cs
> // Schedule it 250 ms in advance so we're not frame-dependant.
> yield return new WaitUntil(() => startDSPTime - AudioSettings.dspTime <= 0.025);
> audioSource.PlayScheduled(startDSPTime);
> ```

That's what I would suggest. You should be able to go well below 250ms before it becomes unreliable.

Reply:
I can't decide of a value there, ideally as low as possible while still being more than a frame (since `WaitUntil` is at least going to skip a frame).